peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+
+However ``Annotated[]`` forms need not preserve their metadata at
+typechecking-time:
+
+::
+
+   count: PositiveInt = 1
+   assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+
+In particular when an ``Annotated[]`` argument is passed to a ``TypeExpr[]``
+parameter constraining a type variable that is also used by a ``TypeIs[]``
+or ``TypeGuard[]``, only the first type argument to ``Annotated[]`` must be
+maintained in the type inferred by a type checker. The second metadata
+argument to ``Annotated[]`` does need not be maintained:
+
+::
+
+   # Similar to isassignable(), but accepts Annotated[] forms describing constraints
+   def ismatch[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]: ...
+
+   count: int | str = -1
+   if ismatch(count, PositiveInt):
+       assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+   else:
+       assert_type(count, int | str)
+
+The above behavior of stripping the metadata argument to ``Annotated[]`` is
+intended to be consistent with the current behavior of static type
+checkers. Future PEPs may be interested in altering the behavior to
+preserve the metadata argument to ``Annotated[]``, but that is out of scope
+of this PEP.
+
+
+Interactions with isinstance() and issubclass()
+-----------------------------------------------
+
+The ``TypeExpr`` special form is not a ``type`` at runtime. It is
+however an ``object``:
+
+::
+
+   >>> isinstance(TypeExpr, type)
+   False
+
+   >>> isinstance(TypeExpr, object)
+   True
+
+The ``TypeExpr`` special form cannot be used as the ``type`` argument to
+``isinstance``:
+
+::
+
+   >>> isinstance(str, TypeExpr)
+   TypeError: typing.TypeExpr cannot be used with isinstance()
+
+   >>> isinstance(str, TypeExpr[str])
+   TypeError: isinstance() argument 2 cannot be a parameterized generic
+
+The ``TypeExpr`` special form cannot be used as any argument to
+``issubclass``:
+
+::
+
+   >>> issubclass(TypeExpr, object)
+   TypeError: issubclass() arg 1 must be a class
+
+   >>> issubclass(object, TypeExpr)
+   TypeError: typing.TypeExpr cannot be used with issubclass()
+
+
+Affected signatures in the standard library
+-------------------------------------------
+
+Changed signatures
+''''''''''''''''''
+
+The following signatures related to type expressions introduce
+``TypeExpr`` where previously ``object`` existed:
+
+-  ``typing.cast(typ: TypeExpr[T], val: object) -> T``
+-  ``typing.assert_type(val: object, typ: TypeExpr, /) -> None``
+
+
+Unchanged signatures
+''''''''''''''''''''
+
+The following signatures related to annotation expressions continue to
+use ``object`` and remain unchanged:
+
+-  ``typing.get_origin(tp: object) -> object``
+-  ``typing.get_args(tp: object) -> tuple[object, ...]``
+
+The following signatures related to class objects continue to use
+``type`` and remain unchanged:
+
+-  ``builtins.isinstance(object: object, classinfo: type | tuple[type, ...]) -> bool``
```

Comment:
This is not precisely the signature of `isinstance()` (see typeshed for the full version). I'm not sure we need to spell out the signature at all; we can merely list the name of the functions.